# Backport of KhronosGroup/SPIRV-Cross PR #2704 (open at review time).
# Author: Zheorgyan; head 51648cae36ebcb24442333ea3dd074b0c4ea66ae.
# Base/MoltenVK dependency pin: aa217aeb6c9f0ace7a0ab233b28807edf45eb165.
# https://github.com/KhronosGroup/SPIRV-Cross/pull/2704
# Preserve signed zero in precise scalar/vector/matrix FMA helpers.
diff --git a/reference/opt/shaders-msl/vert/float-math.invariant-float-math.vert b/reference/opt/shaders-msl/vert/float-math.invariant-float-math.vert
--- a/reference/opt/shaders-msl/vert/float-math.invariant-float-math.vert
+++ b/reference/opt/shaders-msl/vert/float-math.invariant-float-math.vert
@@ -47,13 +47,13 @@ struct spvUnsafeArray
 template<typename T>
 [[clang::optnone]] T spvFMul(T l, T r)
 {
-    return fma(l, r, T(0));
+    return fma(l, r, T(-0.0));
 }
 
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Cols> spvFMulVectorMatrix(vec<T, Rows> v, matrix<T, Cols, Rows> m)
 {
-    vec<T, Cols> res = vec<T, Cols>(0);
+    vec<T, Cols> res = vec<T, Cols>(T(-0.0));
     for (uint i = Rows; i > 0; --i)
     {
         vec<T, Cols> tmp(0);
@@ -69,7 +69,7 @@ template<typename T, int Cols, int Rows>
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Rows> spvFMulMatrixVector(matrix<T, Cols, Rows> m, vec<T, Cols> v)
 {
-    vec<T, Rows> res = vec<T, Rows>(0);
+    vec<T, Rows> res = vec<T, Rows>(T(-0.0));
     for (uint i = Cols; i > 0; --i)
     {
         res = fma(m[i - 1], vec<T, Rows>(v[i - 1]), res);
@@ -84,7 +84,7 @@ template<typename T, int LCols, int LRows, int RCols, int RRows>
     matrix<T, RCols, LRows> res;
     for (uint i = 0; i < RCols; i++)
     {
-        vec<T, LRows> tmp(0);
+        vec<T, LRows> tmp(T(-0.0));
         for (uint j = 0; j < LCols; j++)
         {
             tmp = fma(vec<T, LRows>(r[i][j]), l[j], tmp);
diff --git a/reference/opt/shaders-msl/vert/no-contraction.vert b/reference/opt/shaders-msl/vert/no-contraction.vert
--- a/reference/opt/shaders-msl/vert/no-contraction.vert
+++ b/reference/opt/shaders-msl/vert/no-contraction.vert
@@ -8,13 +8,13 @@ using namespace metal;
 template<typename T>
 [[clang::optnone]] T spvFMul(T l, T r)
 {
-    return fma(l, r, T(0));
+    return fma(l, r, T(-0.0));
 }
 
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Cols> spvFMulVectorMatrix(vec<T, Rows> v, matrix<T, Cols, Rows> m)
 {
-    vec<T, Cols> res = vec<T, Cols>(0);
+    vec<T, Cols> res = vec<T, Cols>(T(-0.0));
     for (uint i = Rows; i > 0; --i)
     {
         vec<T, Cols> tmp(0);
@@ -30,7 +30,7 @@ template<typename T, int Cols, int Rows>
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Rows> spvFMulMatrixVector(matrix<T, Cols, Rows> m, vec<T, Cols> v)
 {
-    vec<T, Rows> res = vec<T, Rows>(0);
+    vec<T, Rows> res = vec<T, Rows>(T(-0.0));
     for (uint i = Cols; i > 0; --i)
     {
         res = fma(m[i - 1], vec<T, Rows>(v[i - 1]), res);
@@ -45,7 +45,7 @@ template<typename T, int LCols, int LRows, int RCols, int RRows>
     matrix<T, RCols, LRows> res;
     for (uint i = 0; i < RCols; i++)
     {
-        vec<T, LRows> tmp(0);
+        vec<T, LRows> tmp(T(-0.0));
         for (uint j = 0; j < LCols; j++)
         {
             tmp = fma(vec<T, LRows>(r[i][j]), l[j], tmp);
diff --git a/reference/shaders-msl-no-opt/asm/comp/compilation-loop-edge-case.spv16.asm.comp b/reference/shaders-msl-no-opt/asm/comp/compilation-loop-edge-case.spv16.asm.comp
--- a/reference/shaders-msl-no-opt/asm/comp/compilation-loop-edge-case.spv16.asm.comp
+++ b/reference/shaders-msl-no-opt/asm/comp/compilation-loop-edge-case.spv16.asm.comp
@@ -8,13 +8,13 @@ using namespace metal;
 template<typename T>
 [[clang::optnone]] T spvFMul(T l, T r)
 {
-    return fma(l, r, T(0));
+    return fma(l, r, T(-0.0));
 }
 
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Cols> spvFMulVectorMatrix(vec<T, Rows> v, matrix<T, Cols, Rows> m)
 {
-    vec<T, Cols> res = vec<T, Cols>(0);
+    vec<T, Cols> res = vec<T, Cols>(T(-0.0));
     for (uint i = Rows; i > 0; --i)
     {
         vec<T, Cols> tmp(0);
@@ -30,7 +30,7 @@ template<typename T, int Cols, int Rows>
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Rows> spvFMulMatrixVector(matrix<T, Cols, Rows> m, vec<T, Cols> v)
 {
-    vec<T, Rows> res = vec<T, Rows>(0);
+    vec<T, Rows> res = vec<T, Rows>(T(-0.0));
     for (uint i = Cols; i > 0; --i)
     {
         res = fma(m[i - 1], vec<T, Rows>(v[i - 1]), res);
@@ -45,7 +45,7 @@ template<typename T, int LCols, int LRows, int RCols, int RRows>
     matrix<T, RCols, LRows> res;
     for (uint i = 0; i < RCols; i++)
     {
-        vec<T, LRows> tmp(0);
+        vec<T, LRows> tmp(T(-0.0));
         for (uint j = 0; j < LCols; j++)
         {
             tmp = fma(vec<T, LRows>(r[i][j]), l[j], tmp);
diff --git a/reference/shaders-msl-no-opt/comp/precise-non-square-matrix.comp b/reference/shaders-msl-no-opt/comp/precise-non-square-matrix.comp
--- a/reference/shaders-msl-no-opt/comp/precise-non-square-matrix.comp
+++ b/reference/shaders-msl-no-opt/comp/precise-non-square-matrix.comp
@@ -8,13 +8,13 @@ using namespace metal;
 template<typename T>
 [[clang::optnone]] T spvFMul(T l, T r)
 {
-    return fma(l, r, T(0));
+    return fma(l, r, T(-0.0));
 }
 
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Cols> spvFMulVectorMatrix(vec<T, Rows> v, matrix<T, Cols, Rows> m)
 {
-    vec<T, Cols> res = vec<T, Cols>(0);
+    vec<T, Cols> res = vec<T, Cols>(T(-0.0));
     for (uint i = Rows; i > 0; --i)
     {
         vec<T, Cols> tmp(0);
@@ -30,7 +30,7 @@ template<typename T, int Cols, int Rows>
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Rows> spvFMulMatrixVector(matrix<T, Cols, Rows> m, vec<T, Cols> v)
 {
-    vec<T, Rows> res = vec<T, Rows>(0);
+    vec<T, Rows> res = vec<T, Rows>(T(-0.0));
     for (uint i = Cols; i > 0; --i)
     {
         res = fma(m[i - 1], vec<T, Rows>(v[i - 1]), res);
@@ -45,7 +45,7 @@ template<typename T, int LCols, int LRows, int RCols, int RRows>
     matrix<T, RCols, LRows> res;
     for (uint i = 0; i < RCols; i++)
     {
-        vec<T, LRows> tmp(0);
+        vec<T, LRows> tmp(T(-0.0));
         for (uint j = 0; j < LCols; j++)
         {
             tmp = fma(vec<T, LRows>(r[i][j]), l[j], tmp);
diff --git a/reference/shaders-msl/vert/float-math.invariant-float-math.vert b/reference/shaders-msl/vert/float-math.invariant-float-math.vert
--- a/reference/shaders-msl/vert/float-math.invariant-float-math.vert
+++ b/reference/shaders-msl/vert/float-math.invariant-float-math.vert
@@ -47,13 +47,13 @@ struct spvUnsafeArray
 template<typename T>
 [[clang::optnone]] T spvFMul(T l, T r)
 {
-    return fma(l, r, T(0));
+    return fma(l, r, T(-0.0));
 }
 
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Cols> spvFMulVectorMatrix(vec<T, Rows> v, matrix<T, Cols, Rows> m)
 {
-    vec<T, Cols> res = vec<T, Cols>(0);
+    vec<T, Cols> res = vec<T, Cols>(T(-0.0));
     for (uint i = Rows; i > 0; --i)
     {
         vec<T, Cols> tmp(0);
@@ -69,7 +69,7 @@ template<typename T, int Cols, int Rows>
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Rows> spvFMulMatrixVector(matrix<T, Cols, Rows> m, vec<T, Cols> v)
 {
-    vec<T, Rows> res = vec<T, Rows>(0);
+    vec<T, Rows> res = vec<T, Rows>(T(-0.0));
     for (uint i = Cols; i > 0; --i)
     {
         res = fma(m[i - 1], vec<T, Rows>(v[i - 1]), res);
@@ -84,7 +84,7 @@ template<typename T, int LCols, int LRows, int RCols, int RRows>
     matrix<T, RCols, LRows> res;
     for (uint i = 0; i < RCols; i++)
     {
-        vec<T, LRows> tmp(0);
+        vec<T, LRows> tmp(T(-0.0));
         for (uint j = 0; j < LCols; j++)
         {
             tmp = fma(vec<T, LRows>(r[i][j]), l[j], tmp);
diff --git a/reference/shaders-msl/vert/no-contraction.vert b/reference/shaders-msl/vert/no-contraction.vert
--- a/reference/shaders-msl/vert/no-contraction.vert
+++ b/reference/shaders-msl/vert/no-contraction.vert
@@ -8,13 +8,13 @@ using namespace metal;
 template<typename T>
 [[clang::optnone]] T spvFMul(T l, T r)
 {
-    return fma(l, r, T(0));
+    return fma(l, r, T(-0.0));
 }
 
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Cols> spvFMulVectorMatrix(vec<T, Rows> v, matrix<T, Cols, Rows> m)
 {
-    vec<T, Cols> res = vec<T, Cols>(0);
+    vec<T, Cols> res = vec<T, Cols>(T(-0.0));
     for (uint i = Rows; i > 0; --i)
     {
         vec<T, Cols> tmp(0);
@@ -30,7 +30,7 @@ template<typename T, int Cols, int Rows>
 template<typename T, int Cols, int Rows>
 [[clang::optnone]] vec<T, Rows> spvFMulMatrixVector(matrix<T, Cols, Rows> m, vec<T, Cols> v)
 {
-    vec<T, Rows> res = vec<T, Rows>(0);
+    vec<T, Rows> res = vec<T, Rows>(T(-0.0));
     for (uint i = Cols; i > 0; --i)
     {
         res = fma(m[i - 1], vec<T, Rows>(v[i - 1]), res);
@@ -45,7 +45,7 @@ template<typename T, int LCols, int LRows, int RCols, int RRows>
     matrix<T, RCols, LRows> res;
     for (uint i = 0; i < RCols; i++)
     {
-        vec<T, LRows> tmp(0);
+        vec<T, LRows> tmp(T(-0.0));
         for (uint j = 0; j < LCols; j++)
         {
             tmp = fma(vec<T, LRows>(r[i][j]), l[j], tmp);
diff --git a/spirv_msl.cpp b/spirv_msl.cpp
--- a/spirv_msl.cpp
+++ b/spirv_msl.cpp
@@ -6447,14 +6447,14 @@ void CompilerMSL::emit_custom_functions()
 			statement("template<typename T>");
 			statement("[[clang::optnone]] T spvFMul(T l, T r)");
 			begin_scope();
-			statement("return fma(l, r, T(0));");
+			statement("return fma(l, r, T(-0.0));");
 			end_scope();
 			statement("");
 
 			statement("template<typename T, int Cols, int Rows>");
 			statement("[[clang::optnone]] vec<T, Cols> spvFMulVectorMatrix(vec<T, Rows> v, matrix<T, Cols, Rows> m)");
 			begin_scope();
-			statement("vec<T, Cols> res = vec<T, Cols>(0);");
+			statement("vec<T, Cols> res = vec<T, Cols>(T(-0.0));");
 			statement("for (uint i = Rows; i > 0; --i)");
 			begin_scope();
 			statement("vec<T, Cols> tmp(0);");
@@ -6471,7 +6471,7 @@ void CompilerMSL::emit_custom_functions()
 			statement("template<typename T, int Cols, int Rows>");
 			statement("[[clang::optnone]] vec<T, Rows> spvFMulMatrixVector(matrix<T, Cols, Rows> m, vec<T, Cols> v)");
 			begin_scope();
-			statement("vec<T, Rows> res = vec<T, Rows>(0);");
+			statement("vec<T, Rows> res = vec<T, Rows>(T(-0.0));");
 			statement("for (uint i = Cols; i > 0; --i)");
 			begin_scope();
 			statement("res = fma(m[i - 1], vec<T, Rows>(v[i - 1]), res);");
@@ -6487,7 +6487,7 @@ void CompilerMSL::emit_custom_functions()
 			statement("matrix<T, RCols, LRows> res;");
 			statement("for (uint i = 0; i < RCols; i++)");
 			begin_scope();
-			statement("vec<T, LRows> tmp(0);");
+			statement("vec<T, LRows> tmp(T(-0.0));");
 			statement("for (uint j = 0; j < LCols; j++)");
 			begin_scope();
 			statement("tmp = fma(vec<T, LRows>(r[i][j]), l[j], tmp);");
